### PR TITLE
TGUI PTL

### PR DESCRIPTION
**Pull request by Sovexe**
## About the PR 
Converts the PTL to a new TGUI based interface

image

images of previous revision
https://user-images.githubusercontent.com/33204415/90702908-6e149c80-e25a-11ea-9da8-6bbad7b338ff.png
https://user-images.githubusercontent.com/33204415/90703002-9f8d6800-e25a-11ea-8426-5c3b0bf066a6.png

## Why's this needed? 
T-G-U-I

## Changelog

```
(u)Sovexe:
(+)Converted PTL to TGUI
```

### CODE REVIEW

File: tgui/packages/tgui/interfaces/PowerTransmissionLaser.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
+import { useBackend } from '../backend';
+import { Box, Button, ColorBox, Divider, Knob, LabeledList, NoticeBox, ProgressBar, Section } from '../components';
+import { Window } from '../layouts';
+
+export const PowerTransmissionLaser = (props, context) => {
+  const { data } = useBackend(context);
+  // Extract `health` and `color` variables from the `data` object.
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/PowerTransmissionLaser.js:7]: 👀

Comment (Sovexe) [reply tgui/packages/tgui/interfaces/PowerTransmissionLaser.js]: removed

**Comment by Sovexe** [comment]
I've redesigned the UI into an experience that I think should look and feel a lot cleaner.

Additionally I've offloaded the unit notations to JS rather than repeated proc calls in DM.

image

File: code/modules/power/pt_laser.dm [hunk 2]
```diff
@@ -339,158 +337,96 @@
...
+		ui = new(user, src, "PowerTransmissionLaser")
+		ui.open()
+
+/obj/machinery/power/pt_laser/ui_data(mob/user)
+	var/list/data = list()
+	data["capacity"] = src.capacity
+	data["charge"] = src.charge
+	data["chargingEnabled"] = src.charging
+	data["excessPower"] = src.excess
+	data["gridLoad"] = src.terminal.powernet.load
```
Comment (mordent-goonstation) [inline code/modules/power/pt_laser.dm:362]: Consider `?.` here, if for whatever dumb reason it's missing a terminal.

File: code/modules/power/pt_laser.dm [hunk 3]
```diff
@@ -339,158 +337,96 @@
...
+	data["inputNumber"] = src.input_number
+	data["isCharging"] = src.is_charging
+	data["isFiring"] = src.firing
+	data["laserEnabled"] = src.online
+	data["lifetimeEarnings"] = src.lifetime_earnings
+	data["name"] = src.name
+	data["outputLevel"] = src.output
+	data["outputMultiplier"] = src.output_multi
+	data["outputNumber"] = src.output_number
+	data["totalGridPower"] = src.terminal.powernet.avail
```
Comment (mordent-goonstation) [inline code/modules/power/pt_laser.dm:374]: Consider `?.` here, if for whatever dumb reason it's missing a terminal.

File: code/modules/power/pt_laser.dm [hunk 4]
```diff
@@ -339,158 +337,96 @@
...
+			src.chargelevel = src.input_number * src.input_multi
+			. = TRUE
+		if("inputGW")
+			src.input_multi = 1 GIGA WATT
+			src.chargelevel = src.input_number * src.input_multi
+			. = TRUE
+		if("inputTW")
+			src.input_multi = 1 TERA WATT
+			src.chargelevel = src.input_number * src.input_multi
+			. = TRUE
```
Comment (mordent-goonstation) [inline code/modules/power/pt_laser.dm:408]: Future reference, could do as a single action with params, but I get that you'd then have to convert the param into a number anyway so NBD here. It's neat enough.

File: tgui/packages/tgui/interfaces/PowerTransmissionLaser.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+import { useBackend } from '../backend';
+import { Box, Button, ColorBox, Divider, Knob, LabeledList, NoticeBox, ProgressBar, Section } from '../components';
+import { formatMoney, formatPower, formatSiUnit } from '../format';
+import { Window } from '../layouts';
+
+export const PowerTransmissionLaser = (props, context) => {
+  const { data } = useBackend(context);
+  const {
+    lifetimeEarnings,
+    name,
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/PowerTransmissionLaser.js:17]: Fine I think in this case, just noting that sensible defaults for this sort of thing are probably good.
```suggestion
    name = 'Power Transmission Laser',
```

File: tgui/packages/tgui/interfaces/PowerTransmissionLaser.js [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+      </Box>
+    </Section>
+  );
+};
+
+const OutputControls = (props, context) => {
+  const { act, data } = useBackend(context);
+  const {
+    isFiring,
+    laserEnabled,
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/PowerTransmissionLaser.js:158]: Mix of variable naming convention here (i.e. one using "is", one not) seems meh. Not requiring you change it, just highlighting code style stuff that we'll never be able to lint so might as well spread the knowledge.

I have no problem with `firing` as a key on `data`.

**Review by mordent-goonstation — CHANGES_REQUESTED** [review]
Only change I think I'm "requiring" is the `?.` ones.

File: code/modules/power/pt_laser.dm (hunk 2 above)
Comment (Sovexe) [reply code/modules/power/pt_laser.dm]: added

File: code/modules/power/pt_laser.dm (hunk 3 above)
Comment (Sovexe) [reply code/modules/power/pt_laser.dm]: added

File: tgui/packages/tgui/interfaces/PowerTransmissionLaser.js (hunk 6 above)
Comment (Sovexe) [reply tgui/packages/tgui/interfaces/PowerTransmissionLaser.js]: vars have been renamed to `isLaserEnabled` and `isChargingEnabled` to match `isFiring` and `isCharging`

File: tgui/packages/tgui/interfaces/PowerTransmissionLaser.js [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    <Window
+      title={name}
+      width="310"
+      height="485">
+      <Window.Content>
+        <Status />
+        <InputControls />
+        <OutputControls />
+        <NoticeBox success>
+          Earned Credits : {formatMoney(lifetimeEarnings)}
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/PowerTransmissionLaser.js:29]: does this not put commas between the 3-unit-seperation? doesn't look like it on the image

**Review by ZeWaka — APPROVED** [review]

File: tgui/packages/tgui/interfaces/PowerTransmissionLaser.js (hunk 7 above)
Comment (Sovexe) [reply tgui/packages/tgui/interfaces/PowerTransmissionLaser.js:29]: It is just spaces, we can change it if you want

**Pull request merged**